### PR TITLE
feat(provider): INT-4969 add 3DS2 flow under experiment for Opayo

### PR DESCRIPTION
**Pull request by JoseAlvarezHdz**
## What? [INT-4969](https://jira.bigcommerce.com/browse/INT-4969)

add 3DS2 flow under experiment for Opayo

## Why?

In order to authenticate shoppers via 3DS2 or 3DS1 when they checkout using Opayo Direct so that can be adhere to PSD2 regulation

### 3DS v2
Untitled_ Jan 28, 2022 2_02 PM

### fallback to 3DS v1
Untitled_ Jan 28, 2022 2_06 PM

## Depends on

[BIGPAY-CLIENT-JS](https://github.com/bigcommerce/bigpay-client-js/pull/125)
[BCAPP](https://github.com/bigcommerce/bigcommerce/pull/44710)

## Dependency of

[BIGPAY](https://github.com/bigcommerce/bigpay/pull/4909)

@bigcommerce/checkout @bigcommerce/payments

### CODE REVIEW

File: src/payment/strategies/sage-pay/sage-pay-payment-strategy.ts
Comment (mauricio-sg) [inline src/payment/strategies/sage-pay/sage-pay-payment-strategy.ts:26]: You can just use `!payment` 🤷🏻‍♂️

File: src/payment/strategies/sage-pay/sage-pay-payment-strategy.ts
Comment (mauricio-sg) [inline src/payment/strategies/sage-pay/sage-pay-payment-strategy.ts:46]: You can just use `!error.body.three_ds_result.merchant_data` 🤷🏻‍♂️

File: src/payment/strategies/sage-pay/sage-pay-payment-strategy.ts
Comment (mauricio-sg) [inline src/payment/strategies/sage-pay/sage-pay-payment-strategy.ts:78]:
```suggestion
        return this._store.getState().config.getStoreConfigOrThrow()
            .checkoutSettings.features['INT-4994.Opayo_3DS2'] === true;
```

**Review by mauricio-sg — APPROVED** [review]
LGTM, is up to you to follow my suggestions 🍹

File: src/payment/strategies/sage-pay/sage-pay-payment-strategy.ts
Comment (bc-peng) [reply src/payment/strategies/sage-pay/sage-pay-payment-strategy.ts:26]:
```suggestion
        if (!payload.payment) {
```

**Comment by JoseAlvarezHdz** [comment]
@bc-peng do you mind taking a look again please?

File: src/payment/strategies/sage-pay/sage-pay-payment-strategy.ts
Comment (bc-peng) [inline src/payment/strategies/sage-pay/sage-pay-payment-strategy.ts:76]: Could we put this in one line? I believe your earlier approach is clear and tidy, though spanning across multiple lines. This approach is also good, but I'm concerned that the line break in between might confuse developers.

File: src/payment/strategies/sage-pay/sage-pay-payment-strategy.ts
Comment (bc-peng) [inline src/payment/strategies/sage-pay/sage-pay-payment-strategy.ts:44]: Could we add `type` information to `payload`?

File: src/payment/strategies/sage-pay/sage-pay-payment-strategy.ts
Comment (animesh1987) [inline src/payment/strategies/sage-pay/sage-pay-payment-strategy.ts:38]:
```suggestion
        const { payment } = payload;
        if (!payment) {
            throw new NotInitializedError(NotInitializedErrorType.PaymentNotInitialized);
        }
        
        const { paymentData } = payment;
        
        if (!paymentData) {
            throw new MissingDataError(MissingDataErrorType.MissingPayment);
        }

        if (this._isThreeDSTwoExperimentOn()) {
            paymentData = {
                ....paymentData,
                browser_info: getBrowserInfo(),
            };
        }

```

File: src/payment/strategies/sage-pay/sage-pay-payment-strategy.ts
Comment (animesh1987) [inline src/payment/strategies/sage-pay/sage-pay-payment-strategy.ts:33]: Also where is this used?

File: src/payment/strategies/sage-pay/sagepay.ts
Comment (animesh1987) [inline src/payment/strategies/sage-pay/sagepay.ts:1]: Why do we need to create this interface? Since it's simply equal to `{}`

File: src/payment/strategies/sage-pay/sage-pay-payment-strategy.ts
Comment (JoseAlvarezHdz) [reply src/payment/strategies/sage-pay/sage-pay-payment-strategy.ts:38]: I think this wont work as we are adding parameters to the payload object, we will have to do something like this
 
        const { payment } = payload;
        if (!payment) {
            throw new NotInitializedError(NotInitializedErrorType.PaymentNotInitialized);
        }

        const { paymentData } = payment;

        if (!paymentData) {
            throw new MissingDataError(MissingDataErrorType.MissingPayment);
        }

        if (this._isThreeDSTwoExperimentOn()) {
            const newPaymentData = {
                ...paymentData,
                browser_info: getBrowserInfo(),
            };

            if ( payload.payment ) {
                payload.payment.paymentData = newPaymentData;
            }
        }
        
and the two ifs will be useless

File: src/payment/strategies/sage-pay/sage-pay-payment-strategy.ts
Comment (JoseAlvarezHdz) [reply src/payment/strategies/sage-pay/sage-pay-payment-strategy.ts:33]: It is used here `return super.execute(payload, options)` so all the browser_info is sent to [bigpay-client-js ](https://github.com/bigcommerce/bigpay-client-js/pull/125/files#diff-3da05841578808d50f2a80c3653813d5686ef321688400f603b71809929e26b3R77) and then its sent to bigpay

**Comment by JoseAlvarezHdz** [comment]
@animesh1987 @bc-peng  would you mind taking a look again please!

File: src/payment/strategies/sage-pay/sagepay.ts
Comment (animesh1987) [reply src/payment/strategies/sage-pay/sagepay.ts:1]: I think you can rid of the type here, since it's actually misleading

File: src/payment/strategies/sage-pay/sage-pay-payment-strategy.ts
Comment (animesh1987) [reply src/payment/strategies/sage-pay/sage-pay-payment-strategy.ts:38]: Yes can we do the above then, the ifs are not useless since they act as protection in case the value is null/undefined. Which can happen.

Comment (bc-peng) [reply src/payment/strategies/sage-pay/sage-pay-payment-strategy.ts:38]: The end-use facing rerror messages for `NotInitializedError(NotInitializedErrorType.PaymentNotInitialized)` and `MissingDataError(MissingDataErrorType.MissingPayment)` are different. 

In my opinion, it is a better experience to provide accurate error information.

Besides this, all look good to me.

**Comment by JoseAlvarezHdz** [comment]
@animesh1987 @bc-peng would you mind taking a look again please!

File: src/payment/strategies/sage-pay/sage-pay-payment-strategy.ts
Comment (JoseAlvarezHdz) [reply src/payment/strategies/sage-pay/sage-pay-payment-strategy.ts]: makes senses to me, thanks! changes are addressed

**Review by bc-peng — APPROVED** [review]
Thank you very much.  Please wait for Animesh's feedback before merge. 🙏🏻

File: src/payment/strategies/sage-pay/sage-pay-payment-strategy.ts
Comment (animesh1987) [inline src/payment/strategies/sage-pay/sage-pay-payment-strategy.ts:55]: Do we need to declare a type here? You can leave it as `let payload;` and let typescript compiler infer it.

Comment (JoseAlvarezHdz) [reply src/payment/strategies/sage-pay/sage-pay-payment-strategy.ts]: sure thing!

**Comment by JoseAlvarezHdz** [comment]
@animesh1987  do you mind taking a look again please!

**Review by animesh1987 — APPROVED** [review]
Nice work @JoseAlvarezHdz . Thanks for the patience ☺️

**Pull request merged**
